File: returned/python/today15/largest_number.py
# https://leetcode.com/problems/largest-number/
from typing import List

class Solution:
    def largestNumber(self, nums: List[int]) -> str:
        numStrs = [str(n) for n in nums]
        numStrs = self._mergeSort(numStrs)
        joined = "".join(numStrs)
        trimmed = self._trimLeadingZeros(joined)
        return trimmed
    
    def _trimLeadingZeros(self, s: str) -> str:
        i = 0
        while s[i] == '0' and i < len(s) - 1:
            i += 1
        return s[i:]

    def _mergeSort(self, numStrs: List[str]) -> List[str]:
        def merge(left: List[str], right: List[str]) -> List[str]:
            result = []
            i = 0
            j = 0
            while i < len(left) and j < len(right):
                if left[i] + right[j] > right[j] + left[i]:
                    result.append(left[i])
                    i += 1
                else:
                    result.append(right[j])
                    j += 1
            while i < len(left):
                result.append(left[i])
                i += 1
            while j < len(right):
                result.append(right[j])
                j += 1
            return result

        if len(numStrs) <= 1:
            return numStrs
        
        mid = len(numStrs) // 2
        left = self._mergeSort(numStrs[:mid])
        right = self._mergeSort(numStrs[mid:])
        return merge(left, right)

File: csharp/Leet/Old/LargestNumber.cs
// https://leetcode.com/problems/largest-number/
namespace Old.LargestNumber
{
    public class Solution
    {
        public string LargestNumber(int[] nums)
        {
            var numStrings = new string[nums.Length];
            for (var i = 0; i < nums.Length; i++)
            {
                numStrings[i] = nums[i].ToString();
            }
            numStrings = MergeSort(numStrings);

            var result = "";
            foreach (var s in numStrings)
            {
                result += s;
            }

            var leadingZeroIndex = 0;
            for (var i = 0; i < result.Length - 1; i++)
            {
                if (result[i] == '0') leadingZeroIndex = i + 1;
                else break;
            }

            return result[leadingZeroIndex..];
        }

        public string[] MergeSort(string[] nums)
        {
            if (nums.Length == 1) return nums;
            var mid = (Index)Math.Floor(nums.Length / 2d);
            var left = MergeSort(nums[0..mid]);
            var right = MergeSort(nums[mid..]);
            return Merge(left, right);
        }

        public string[] Merge(string[] left, string[] right)
        {
            var result = new string[left.Length + right.Length];
            int i = 0, l = 0, r = 0;

            while (l < left.Length && r < right.Length)
            {
                if (CompareStrings(left[l], right[r]))
                    result[i] = right[r++];
                else
                    result[i] = left[l++];
                i++;
            }

            while (l < left.Length)
            {
                result[i] = left[l++];
                i++;
            }

            while (r < right.Length)
            {
                result[i] = right[r++];
                i++;
            }

            return result;
        }

        public bool CompareStrings(string a, string b)
        {
            var ab = a + b;
            var ba = b + a;
            for (var i = 0; i < a.Length + b.Length; i++)
            {
                var compared = ab[i].CompareTo(ba[i]);
                if (compared < 0) return true;
                else if (compared > 0) return false;
            }
            return true;
        }
    }
}
